lntest/itest/rest_api.go

```diff
@@ -0,0 +1,352 @@
+	}
+	restClient = &http.Client{
+		Transport: insecureTransport,
+	}
+	jsonMarshaler = &jsonpb.Marshaler{
+		EmitDefaults: true,
+		OrigName:     true,
+		Indent:       "    ",
+	}
+	urlEnc          = base64.URLEncoding
+	webSocketDialer = &websocket.Dialer{
+		HandshakeTimeout: 45 * time.Second,
+		TLSClientConfig:  insecureTransport.TLSClientConfig,
+	}
+	resultPattern = regexp.MustCompile("{\"result\":(.*)}")
+)
+
+// testRestApi tests that the most important features of the REST API work
+// correctly.
+func testRestApi(net *lntest.NetworkHarness, ht *harnessTest) {
+	testCases := []struct {
+		name string
+		run  func(*testing.T, *lntest.HarnessNode, *lntest.HarnessNode)
+	}{{
+		name: "simple GET",
+		run: func(t *testing.T, a, b *lntest.HarnessNode) {
+			// Check that the parsing into the response prot message
+			// works.
+			resp := &lnrpc.GetInfoResponse{}
+			err := invokeGET(a, "/v1/getinfo", resp)
+			require.Nil(t, err, "getinfo")
+			assert.Equal(t, "#3399ff", resp.Color, "node color")
+
+			// Make sure we get the correct field names (snake
+			// case).
+			resp2, err := makeRequest(a, "/v1/getinfo", "GET", nil)
+			require.Nil(t, err, "getinfo")
+			assert.Contains(
+				t, string(resp2), "best_header_timestamp",
+				"getinfo",
+			)
+		},
+	}, {
+		name: "simple POST and GET with query param",
+		run: func(t *testing.T, a, b *lntest.HarnessNode) {
+			// Add an invoice, testing POST in the process.
+			req := &lnrpc.Invoice{Value: 1234}
+			resp := &lnrpc.AddInvoiceResponse{}
+			err := invokePOST(a, "/v1/invoices", req, resp)
+			require.Nil(t, err, "add invoice")
+			assert.Equal(t, 32, len(resp.RHash), "invoice rhash")
+
+			// Make sure we can call a GET endpoint with a hex
+			// encoded URL part.
+			url := fmt.Sprintf("/v1/invoice/%x", resp.RHash)
+			resp2 := &lnrpc.Invoice{}
+			err = invokeGET(a, url, resp2)
+			require.Nil(t, err, "query invoice")
+			assert.Equal(t, int64(1234), resp2.Value, "invoice amt")
+		},
+	}, {
+		name: "GET with base64 encoded byte slice in path",
+		run: func(t *testing.T, a, b *lntest.HarnessNode) {
+			url := "/v2/router/mc/probability/%s/%s/%d"
+			url = fmt.Sprintf(
+				url, urlEnc.EncodeToString(a.PubKey[:]),
+				urlEnc.EncodeToString(b.PubKey[:]), 1234,
+			)
+			resp := &routerrpc.QueryProbabilityResponse{}
+			err := invokeGET(a, url, resp)
+			require.Nil(t, err, "query probability")
+			assert.Greater(t, resp.Probability, 0.5, "probability")
+		},
+	}, {
+		name: "GET with map type query param",
+		run: func(t *testing.T, a, b *lntest.HarnessNode) {
+			// Get a new wallet address from Alice.
+			ctxb := context.Background()
+			newAddrReq := &lnrpc.NewAddressRequest{
+				Type: lnrpc.AddressType_WITNESS_PUBKEY_HASH,
+			}
+			addrRes, err := a.NewAddress(ctxb, newAddrReq)
+			require.Nil(t, err, "get address")
+
+			// Create the full URL with the map query param.
+			url := "/v1/transactions/fee?target_conf=%d&" +
+				"AddrToAmount[%s]=%d"
```

Comment:
Oh wow, totally didn't know this was how you specified map params :p

Reply:
Yeah, that's really un-intuitive. We should avoid using GET for any messages that contain map types. I added an explicit comment to the only two RPCs where this syntax is needed.